### PR TITLE
kernel: sem: handle resets with outstanding waiting threads

### PR DESCRIPTION
**Pull request by jharris-intel**
Previously, a k_sem_reset with any outstanding waiting threads would
result in the semaphore in an inconsistent state, with more threads
waiting in the wait_q than the count would indicate.

Explicitly -EAGAIN any waiting threads upon k_sem_reset, to
ensure safety here.

Signed-off-by: James Harris <james.harris@intel.com>

### CODE REVIEW

**Review by andyross — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: kernel/sem.c [hunk 1]
```diff
@@ -163,6 +163,25 @@ int z_impl_k_sem_take(struct k_sem *sem, k_timeout_t timeout)
 	return ret;
 }
 
+void z_impl_k_sem_reset(struct k_sem *sem)
+{
+	struct k_thread *thread;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	while (1) {
```
[the remark this answers is not in the document]
Comment (jharris-intel) [reply kernel/sem.c:171]: Good to know, and easy enough to change. Somewhat surprised that isn't caught by checkpatch...

- [ ] Change to `while (true) {`

**Review by dcpleung — APPROVED** [review]

**Comment by nashif** [comment]
adding @jukkar to look into why networking tests are failing...

**Comment by jharris-intel** [comment]
There _are_ a few k_sem_reset calls in various places in the network stack. It's entirely possible something relied on the old (broken) behavior...

Looking at it, we're dying in the `k_sem_reset` call in `check_interface`, with an access violation in `z_handle_obj_poll_events` because  `node` is `NULL` but we're dereferencing `node->next`... which means that `list->next` was `NULL`.

...which likely means we called `k_sem_reset` on a not-properly-initialized semaphore.

...and we totally are:

```c
	if (check_interface(iface) == false) {
		k_sem_init(&counter, 1, UINT_MAX);
```
...where check_interface is the one calling `k_sem_reset`.

Luckily, it's a fairly straightforward fix.

File: subsys/net/lib/config/init.c [hunk 2]
```diff
@@ -32,7 +32,7 @@ extern const struct log_backend *log_backend_net_get(void);
 extern int net_init_clock_via_sntp(void);
 
 static K_SEM_DEFINE(waiter, 0, 1);
-static struct k_sem counter;
+static K_SEM_DEFINE(counter, 0, UINT_MAX);
```
Comment (jharris-intel) [inline subsys/net/lib/config/init.c:35]: Initialize the semaphore to zero available tokens, which should match previous behavior. (...mostly. I'm pretty sure that previously this was "initialized" to a limit of zero, so e.g. `k_sem_give` calls would be silently ignored until the init, but I also suspect that was a bug...).

I don't know if this is the proper proper way to do this here - I'm inclined to say "directly initialize this to 1/UINT_MAX and get rid of the k_sem_init call below", but I don't know the networking subsystem at all.

Comment (jharris-intel) [reply subsys/net/lib/config/init.c:35]: Also, TIL that any polling APIs on a zero-initialized struct will null-pointer-dereference.

**Review by jukkar — APPROVED** [review]

**Pull request merged**